kong/runloop/balancer/ewma.lua

```diff
@@ -0,0 +1,255 @@
+-- ewma balancer algorithm
+--
+-- Original Authors: Shiv Nagarajan & Scott Francis
+-- Accessed: March 12, 2018
+-- Inspiration drawn from:
+-- https://github.com/twitter/finagle/blob/1bc837c4feafc0096e43c0e98516a8e1c50c4421
+--   /finagle-core/src/main/scala/com/twitter/finagle/loadbalancer/PeakEwma.scala
+
+
+local balancers = require "kong.runloop.balancer.balancers"
+
+local ngx_now = ngx.now
+local ngx_log = ngx.log
+local ngx_DEBUG = ngx.DEBUG
+
+local table_insert = table.insert
+
+local DECAY_TIME = 10 -- this value is in seconds
+local PICK_SET_SIZE = 2
+
+local ewma = {}
+ewma.__index = ewma
+
+local function decay_ewma(ewma, last_touched_at, rtt, now)
+    local td = now - last_touched_at
+    td = (td > 0) and td or 0
+    local weight = math.exp(-td/DECAY_TIME)
+  
+    ewma = ewma * weight + rtt * (1.0 - weight)
+    return ewma
+end
+
+
+-- slow_start_ewma is something we use to avoid sending too many requests
+-- to the newly introduced endpoints. We currently use average ewma values
+-- of existing endpoints.
+local function calculate_slow_start_ewma(self)
+    local total_ewma = 0
+    self.address_count = 0
+  
+    for _, target in ipairs(self.balancer.targets) do
+        for _, address in ipairs(target.addresses) do
+            if address.available then
+              local ewma = self.ewma[address] or 0
+              self.address_count = self.address_count + 1
+              total_ewma = total_ewma + ewma
+            end
+        end
+    end
+  
+    if self.address_count == 0 then
+      ngx_log(ngx_DEBUG, "no ewma value exists for the endpoints")
+      return nil
+    end
+  
+    return total_ewma / self.address_count
+end
+
+
+function ewma:afterHostUpdate()
+  local new_addresses = {}
+  
+  for _, target in ipairs(self.balancer.targets) do
+    for _, address in ipairs(target.addresses) do
+      if address.available then
+        new_addresses[address] = true
+      end
+    end
+  end
+
+  for address, _ in pairs(self.ewma) do
+    if not new_addresses[address] then
+      self.ewma[address] = nil
+      self.ewma_last_touched_at[address] = nil
+    end
+  end
+  
+  local slow_start_ewma = calculate_slow_start_ewma(self)
+  if slow_start_ewma ~= nil then
+    local now = ngx_now()
+    for address, _ in pairs(new_addresses) do
+      self.ewma[address] = slow_start_ewma
+      self.ewma_last_touched_at[address] = now
+    end
+  end
+end
+
+
+local function get_or_update_ewma(self, address, rtt, update)
+  local ewma = self.ewma[address] or 0
+  local now = ngx_now()
+  local last_touched_at = self.ewma_last_touched_at[address] or 0
+  ewma = decay_ewma(ewma, last_touched_at, rtt, now)
+  if not update then
+    return ewma
+  end
+
+  self.ewma_last_touched_at[address] = now
+  self.ewma[address] = ewma
+  return ewma
+end
+
+
+function ewma:afterBalance(ctx, handle)
+  local response_time = ngx.var.upstream_response_time or 0
+  local connect_time = ngx.var.upstream_connect_time or 0
```

Comment:
same as above should use `tonumber`. 

Just curious would it ever be possible for `upstream_response_time` or `upstream_connect_time` to not be set?